# Putting together all the data frames and cleaning the data

Import of libraries

In [1]:
import pandas as pd
import sys

Importing .py doc to access the needed function

In [2]:
# Printing all directories for interpreter to search 

sys.path.append('../')

# Importing module

from src.support import *

Import of the data frames

In [3]:
books1 = pd.read_csv('../data/books_information.csv')
books2 = pd.read_csv('../data/books_information1.csv')
books3 = pd.read_csv('../data/books_information2.csv')
books4 = pd.read_csv('../data/books_information3.csv')

In [4]:
display(books1.head(2))
display(books2.head(2))
display(books3.head(2))
books4.head(2)

Cover  \
0  https://images-na.ssl-images-amazon.com/images...   
1  https://images-na.ssl-images-amazon.com/images...   

                                       Title           Author  \
0                           The Hunger Games  Suzanne Collins   
1  Harry Potter and the Order of the Phoenix     J.K. Rowling   

                                       Genre  Rating  \
0  Young , Adult, Fiction, Fantasy, Dystopia    4.33   
1   Young , Adult, Fiction, Magic, Childrens    4.50   

                       Rating Count                                Year  \
0  8,272,162 ratings207,967 reviews  First published September 14, 2008   
1   3,267,591 ratings60,535 reviews       First published June 21, 2003   

                  Pages  
0  374 pages, Hardcover  
1  912 pages, Paperback

Cover  \
0  https://images-na.ssl-images-amazon.com/images...   
1  https://images-na.ssl-images-amazon.com/images...   

                                      Title        Author  \
0                     To Kill a Mockingbird    Harper Lee   
1  Harry Potter and the Philosopher’s Stone  J.K. Rowling   

                                             Genre  Rating  \
0  Classics, Fiction, Historical , Fiction, School    4.26   
1         Fiction, Young , Adult, Magic, Childrens    4.47   

                       Rating Count                           Year  \
0  5,889,033 ratings114,192 reviews  First published July 11, 1960   
1  9,637,191 ratings155,419 reviews  First published June 26, 1997   

                  Pages  
0  323 pages, Paperback  
1  223 pages, Paperback

Cover  \
0  https://images-na.ssl-images-amazon.com/images...   
1  https://images-na.ssl-images-amazon.com/images...   

                      Title              Author  \
0  A Certain Slant of Light      Laura Whitcomb   
1         Nobody's Princess  Esther M. Friesner   

                                          Genre  Rating  \
0   Young , Adult, Fantasy, Paranormal, Romance    3.84   
1  Historical , Fiction, Young , Adult, Fantasy    3.69   

                  Rating Count                                Year  \
0  19,520 ratings1,986 reviews  First published September 21, 2005   
1  13,268 ratings1,080 reviews      First published April 24, 2007   

                  Pages  
0  282 pages, Paperback  
1  296 pages, Hardcover

Cover  \
0  https://images-na.ssl-images-amazon.com/images...   
1  https://images-na.ssl-images-amazon.com/images...   

                      Title         Author  \
0  Daughter of Smoke & Bone   Laini Taylor   
1                  Delirium  Lauren Oliver   

                                         Genre  Rating  \
0  Fantasy, Young , Adult, Romance, Paranormal    3.99   
1    Young , Adult, Dystopia, Romance, Fantasy    3.95   

                    Rating Count                                Year  \
0  364,585 ratings33,599 reviews  First published September 27, 2011   
1  462,169 ratings28,389 reviews    First published February 3, 2011   

                  Pages  
0  418 pages, Hardcover  
1  441 pages, Paperback

Putting them together

In [5]:
books = pd.concat([books1, books2, books3, books4], axis = 0)
books.head()

Cover  \
0  https://images-na.ssl-images-amazon.com/images...   
1  https://images-na.ssl-images-amazon.com/images...   
2  https://images-na.ssl-images-amazon.com/images...   
3  https://images-na.ssl-images-amazon.com/images...   
4  https://images-na.ssl-images-amazon.com/images...   

                                       Title           Author  \
0                           The Hunger Games  Suzanne Collins   
1  Harry Potter and the Order of the Phoenix     J.K. Rowling   
2                        Pride and Prejudice      Jane Austen   
3                      To Kill a Mockingbird       Harper Lee   
4                             The Book Thief     Markus Zusak   

                                              Genre  Rating  \
0         Young , Adult, Fiction, Fantasy, Dystopia    4.33   
1          Young , Adult, Fiction, Magic, Childrens    4.50   
2  Classics, Fiction, Romance, Historical , Fiction    4.28   
3   Classics, Fiction, Historical , Fiction, School    4.26   
4      Historical , Fiction, Fiction, Young , Adult    4.39   

                       Rating Count                                Year  \
0  8,272,162 ratings207,967 reviews  First published September 14, 2008   
1   3,267,591 ratings60,535 reviews       First published June 21, 2003   
2  4,103,863 ratings108,310 reviews    First published January 28, 1813   
3  5,888,672 ratings114,187 reviews       First published July 11, 1960   
4  2,459,198 ratings140,987 reviews       First published March 1, 2005   

                  Pages  
0  374 pages, Hardcover  
1  912 pages, Paperback  
2  279 pages, Paperback  
3  323 pages, Paperback  
4  592 pages, Hardcover

Exploring the data

In [6]:
# Drop duplicates

books = books.drop_duplicates(subset = 'Title', keep = 'first')

In [7]:
books.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 24736 entries, 0 to 9731
Data columns (total 8 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Cover         24736 non-null  object 
 1   Title         24736 non-null  object 
 2   Author        24736 non-null  object 
 3   Genre         22977 non-null  object 
 4   Rating        24736 non-null  float64
 5   Rating Count  24736 non-null  object 
 6   Year          24736 non-null  object 
 7   Pages         24736 non-null  object 
dtypes: float64(1), object(7)
memory usage: 1.7+ MB


In [8]:
books['Genre'].isnull().sum()

1759

Although the 'Genre' column has 1759 Nan values, I'm not gonna delete those rows because you can still filter out those books through other parameters

## Columns' data cleaning

'Title'

In [9]:
books['Title'] = books['Title'].apply(lambda x: str(x) if not x.isdigit() else f'"{x}"')

'Genre'

In [10]:
# Overview of the data

books['Genre'].head()

0           Young , Adult, Fiction, Fantasy, Dystopia
1            Young , Adult, Fiction, Magic, Childrens
2    Classics, Fiction, Romance, Historical , Fiction
3     Classics, Fiction, Historical , Fiction, School
4        Historical , Fiction, Fiction, Young , Adult
Name: Genre, dtype: object

In [11]:
# Filling Nan values with 'Unknown'

books['Genre'] = books['Genre'].fillna('Unknown')

In [12]:
# Deleting spaces between the end of a word and its following comma

books['Genre'] = books['Genre'].str.replace(' ,', ',')

In [13]:
# Deleting the word 'Genre(s)' from the 'Genre' column

books['Genre'] = books['Genre'].str.replace('Genre(s)?', '', regex = True)

In [14]:
# Deleting duplicate values per row in 'Genre' column calling the function 'delete_duplicates'

books['Genre'] = books['Genre'].apply(remove_duplicates)

In [15]:
# Applying the function to the 'Genre' column

books['Genre'] = books['Genre'].apply(common_genres_cleaning)

In [16]:
# Changing the data type of the 'Genre' column to string

books['Genre'] = books['Genre'].apply(lambda x: ', '.join(x))

In [17]:
# Converting the 'Genre' column to a list and splitting it by commas

books['Genre'] = books['Genre'].str.split(',')

# Applying the function 'create_genre_dict' to the 'Genre' column to create a dictionary

books['Genre'] = books['Genre'].apply(create_genre_dict)

In [18]:
# Checking the result

books['Genre']

0       {'genre_1': 'Young', 'genre_2': 'Adult', 'genr...
1       {'genre_1': 'Young', 'genre_2': 'Adult', 'genr...
2       {'genre_1': 'Classics', 'genre_2': 'Fiction', ...
3       {'genre_1': 'Classics', 'genre_2': 'Fiction', ...
4       {'genre_1': 'Historical', 'genre_2': 'Fiction'...
                              ...                        
9722    {'genre_1': 'Mystery', 'genre_2': 'Fiction', '...
9723                               {'genre_1': 'Unknown'}
9724    {'genre_1': 'Fantasy', 'genre_2': 'Science', '...
9728    {'genre_1': 'Science', 'genre_2': 'Fiction', '...
9731    {'genre_1': 'Science', 'genre_2': 'Fiction', '...
Name: Genre, Length: 24736, dtype: object

'Rating'

In [19]:
# Checking for unique values in the column 'Rating'

books['Rating'].unique()

array([4.33, 4.5 , 4.28, 4.26, 4.39, 3.65, 3.98, 4.61, 4.27, 4.14, 4.12,
       4.31, 4.38, 3.89, 4.23, 4.15, 4.47, 3.93, 4.3 , 4.32, 3.91, 4.06,
       4.2 , 3.69, 3.74, 4.07, 4.01, 3.88, 3.9 , 3.8 , 3.97, 3.99, 4.11,
       4.43, 4.44, 4.19, 4.34, 3.86, 4.62, 3.85, 3.83, 4.24, 4.13, 3.87,
       4.09, 4.58, 4.05, 4.35, 4.08, 3.79, 4.16, 4.1 , 4.36, 4.29, 4.  ,
       4.02, 4.25, 4.56, 3.54, 4.52, 4.03, 3.94, 3.92, 3.61, 3.43, 4.18,
       4.04, 4.84, 4.49, 4.45, 4.42, 4.37, 3.62, 3.95, 3.82, 4.17, 3.7 ,
       3.84, 4.59, 3.75, 4.54, 3.66, 4.64, 4.21, 3.52, 4.22, 4.4 , 3.96,
       3.72, 3.6 , 4.41, 3.76, 3.81, 4.8 , 4.6 , 4.66, 4.55, 4.48, 3.78,
       3.68, 4.57, 3.73, 3.48, 3.58, 4.74, 3.63, 3.77, 3.59, 3.31, 4.46,
       3.57, 4.51, 4.65, 3.5 , 4.76, 4.53, 3.64, 3.46, 3.71, 3.44, 3.39,
       3.51, 3.67, 4.71, 4.72, 3.29, 3.3 , 3.53, 3.56, 3.45, 3.2 , 4.7 ,
       4.73, 4.69, 3.55, 4.63, 3.47, 3.36, 3.09, 4.78, 3.42, 3.22, 3.11,
       3.49, 3.18, 3.26, 3.23, 3.19, 3.1 , 3.21, 3.

In [20]:
# Deleting spaces in the 'Rating' column changing the data type to string and back to float

books['Rating'] = books['Rating'].astype(str).str.strip().astype(float)

'Rating Count'

In [21]:
# Splitting 'Rating Count' column into 'Ratings count' and 'Reviews Count' columns and deleting the original column

books[['Ratings Count', 'Reviews Count']] = books['Rating Count'].str.split('ratings', expand = True)
books = books.drop('Rating Count', axis = 1)

In [22]:
# Deleting 'review(s)' word from 'Reviews Count' column

books['Reviews Count'] = books['Reviews Count'].str.split('reviews', expand = True)[0]
books['Reviews Count'] = books['Reviews Count'].str.split('review', expand = True)[0]    

In [23]:
# Checking the data types of the new columns

print(books['Ratings Count'].isnull().sum())
books['Reviews Count'].isnull().sum()

0


52

In [24]:
# Turning the columns into string type

books['Ratings Count'] = books['Ratings Count'].astype(str)
books['Reviews Count'] = books['Reviews Count'].astype(str)

In [25]:
# Replacing ',' and '\xa0' with '' and extracting the numbers from the columns

books['Ratings Count'] = books['Ratings Count'].str.replace(',', '').str.replace('\xa0', '').str.extract('(\d+)').astype(float).astype(int)
books['Reviews Count'] = books['Reviews Count'].str.replace(',', '').str.replace('\xa0', '').str.extract('(\d+)').astype(float).fillna(0).astype(int)

In [26]:
# Deleting spaces in the 'Ratings Count' column changing the data type to string and back to float

books['Ratings Count'] = books['Ratings Count'].astype(str).str.strip().astype(int)

In [27]:
# Deleting spaces in the 'Reviews Count' column changing the data type to string and back to float

books['Reviews Count'] = books['Reviews Count'].astype(str).str.strip().astype(int)

In [28]:
# Checking the data types of the new columns

books.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 24736 entries, 0 to 9731
Data columns (total 9 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Cover          24736 non-null  object 
 1   Title          24736 non-null  object 
 2   Author         24736 non-null  object 
 3   Genre          24736 non-null  object 
 4   Rating         24736 non-null  float64
 5   Year           24736 non-null  object 
 6   Pages          24736 non-null  object 
 7   Ratings Count  24736 non-null  int64  
 8   Reviews Count  24736 non-null  int64  
dtypes: float64(1), int64(2), object(6)
memory usage: 1.9+ MB


'Year'

In [29]:
# Deleting 'First publised '  and 'Published ' from all the rows in 'Year' column

books['Year'] = books['Year'].str.replace('First published ', '')
books['Year'] = books['Year'].str.replace('Published ', '')

In [30]:
# Deleting the out of range years from the 'Year' column 

books['Year'] = pd.to_datetime(books['Year'], format = '%B %d, %Y', errors = 'coerce')
books = books[books['Year'].notnull()]

In [31]:
# Deleting spaces in the 'Year' column changing the data type to string
books['Year'] = books['Year'].astype(str).str.strip()

In [32]:
# Changing the data type to datetime

books['Year'] = pd.to_datetime(books['Year'], format = '%Y-%m-%d')

In [33]:
# Checking the data types of the 'Year' column

books['Year'].info()

<class 'pandas.core.series.Series'>
Int64Index: 24543 entries, 0 to 9731
Series name: Year
Non-Null Count  Dtype         
--------------  -----         
24543 non-null  datetime64[ns]
dtypes: datetime64[ns](1)
memory usage: 383.5 KB


'Pages'

In [34]:
# Extracting the numbers from 'Pages' column

pages_extracted = books['Pages'].str.extract('(\d+)')

In [35]:
# Filling the NaN values with 0

pages_extracted = pages_extracted.fillna(0)

In [36]:
# Changing the data type of 'Pages' column to int

books['Pages'] = pages_extracted.astype(int)

In [37]:
# Checking the data types of the 'Pages' column

books['Pages'].info()

<class 'pandas.core.series.Series'>
Int64Index: 24543 entries, 0 to 9731
Series name: Pages
Non-Null Count  Dtype
--------------  -----
24543 non-null  int64
dtypes: int64(1)
memory usage: 383.5 KB


Final touches

In [38]:
# Changing names of the columns

books = books.rename(columns = {'Cover': 'cover'})

books = books.rename(columns = {'Title': 'title'})

books = books.rename(columns = {'Author': 'author'})

books = books.rename(columns = {'Genre': 'genre'})

books = books.rename(columns = {'Rating': 'rating'})

books = books.rename(columns = {'Ratings Count': 'rating_count'})

books = books.rename(columns = {'Reviews Count': 'review_count'})

books = books.rename(columns = {'Year': 'publication_date'})

books = books.rename(columns = {'Pages': 'num_page'})

In [39]:
# Organizing the columns

books = books.reindex(columns = ['cover', 'title', 'author', 'genre', 'rating', 'rating_count', 'review_count', 'publication_date', 'num_page'])

In [40]:
# Final check

books.head()

cover  \
0  https://images-na.ssl-images-amazon.com/images...   
1  https://images-na.ssl-images-amazon.com/images...   
2  https://images-na.ssl-images-amazon.com/images...   
3  https://images-na.ssl-images-amazon.com/images...   
4  https://images-na.ssl-images-amazon.com/images...   

                                       title           author  \
0                           The Hunger Games  Suzanne Collins   
1  Harry Potter and the Order of the Phoenix     J.K. Rowling   
2                        Pride and Prejudice      Jane Austen   
3                      To Kill a Mockingbird       Harper Lee   
4                             The Book Thief     Markus Zusak   

                                               genre  rating  rating_count  \
0  {'genre_1': 'Young', 'genre_2': 'Adult', 'genr...    4.33       8272162   
1  {'genre_1': 'Young', 'genre_2': 'Adult', 'genr...    4.50       3267591   
2  {'genre_1': 'Classics', 'genre_2': 'Fiction', ...    4.28       4103863   
3  {'genre_1': 'Classics', 'genre_2': 'Fiction', ...    4.26       5888672   
4  {'genre_1': 'Historical', 'genre_2': 'Fiction'...    4.39       2459198   

   review_count publication_date  num_page  
0        207967       2008-09-14       374  
1         60535       2003-06-21       912  
2        108310       1813-01-28       279  
3        114187       1960-07-11       323  
4        140987       2005-03-01       592

In [41]:
# Info check

books.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 24543 entries, 0 to 9731
Data columns (total 9 columns):
 #   Column            Non-Null Count  Dtype         
---  ------            --------------  -----         
 0   cover             24543 non-null  object        
 1   title             24543 non-null  object        
 2   author            24543 non-null  object        
 3   genre             24543 non-null  object        
 4   rating            24543 non-null  float64       
 5   rating_count      24543 non-null  int64         
 6   review_count      24543 non-null  int64         
 7   publication_date  24543 non-null  datetime64[ns]
 8   num_page          24543 non-null  int64         
dtypes: datetime64[ns](1), float64(1), int64(3), object(4)
memory usage: 1.9+ MB


In [42]:
# Saving the data

books.to_csv('../data/books_cleaned.csv', index = False)